### ISSUE TITLE
Incorrect numbers of warnings makes some tests to fail (test_new_style_middleware_method[pyloop], test_old_style_middleware_class[pyloop], and test_mixed_middleware[pyloop])

### ISSUE THREAD

**Issue by mcepl** [aio-libs/aiohttp#5991]
### Describe the bug

When packaging for openSUSE/Factory five tests fail because numbers of warnings in tests don’t correlate with reality:
```
[  261s] =================================== FAILURES ===================================
[  261s] ______________________ test_old_style_middleware[pyloop] _______________________
[  261s]
[  261s] loop = <_UnixSelectorEventLoop running=False closed=False debug=False>
[  261s] aiohttp_client = <function aiohttp_client.<locals>.go at 0x7fde94fd4280>
[  261s]
[  261s]     async def test_old_style_middleware(loop, aiohttp_client) -> None:
[  261s]         async def handler(request):
[  261s]             return web.Response(body=b"OK")
[  261s]
[  261s]         async def middleware_factory(app, handler):
[  261s]             async def middleware(request):
[  261s]                 resp = await handler(request)
[  261s]                 assert 200 == resp.status
[  261s]                 resp.set_status(201)
[  261s]                 resp.text = resp.text + "[old style middleware]"
[  261s]                 return resp
[  261s]
[  261s]             return middleware
[  261s]
[  261s]         with pytest.warns(DeprecationWarning) as warning_checker:
[  261s]             app = web.Application()
[  261s]             app.middlewares.append(middleware_factory)
[  261s]             app.router.add_route("GET", "/", handler)
[  261s]             client = await aiohttp_client(app)
[  261s]             resp = await client.get("/")
[  261s]             assert 201 == resp.status
[  261s]             txt = await resp.text()
[  261s]             assert "OK[old style middleware]" == txt
[  261s]
[  261s] >       assert len(warning_checker) == 1, f"warning_checker:\n{[str(x) for x in warning_checker.list]}"
[  261s] E       AssertionError: warning_checker:
[  261s] E         ['{message : DeprecationWarning(\'old-style middleware "<function test_old_style_middleware.<locals>.middleware_factory at 0x7fde94fd4160>" deprecated, see #2252\'), category : \'DeprecationWarning\', filename : \'/home/abuild/rpmbuild/BUILDROOT/python-aiohttp-3.7.4-0.x86_64/usr/lib64/python3.9/site-packages/aiohttp/web_app.py\', lineno : 252, line : None}', "{message : DeprecationWarning('The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.'), category : 'DeprecationWarning', filename : '/usr/lib64/python3.9/asyncio/base_events.py', lineno : 1518, line : None}"]
[  261s] E       assert 2 == 1
[  261s] E         +2
[  261s] E         -1
[  261s]
[  261s] tests/test_web_middleware.py:413: AssertionError
[  261s] ________________________ test_mixed_middleware[pyloop] _________________________
[  261s]
[  261s] loop = <_UnixSelectorEventLoop running=False closed=False debug=False>
[  261s] aiohttp_client = <function aiohttp_client.<locals>.go at 0x7fde94fe3280>
[  261s]
[  261s]     async def test_mixed_middleware(loop, aiohttp_client) -> None:
[  261s]         async def handler(request):
[  261s]             return web.Response(body=b"OK")
[  261s]
[  261s]         async def m_old1(app, handler):
[  261s]             async def middleware(request):
[  261s]                 resp = await handler(request)
[  261s]                 resp.text += "[old style 1]"
[  261s]                 return resp
[  261s]
[  261s]             return middleware
[  261s]
[  261s]         @web.middleware
[  261s]         async def m_new1(request, handler):
[  261s]             resp = await handler(request)
[  261s]             resp.text += "[new style 1]"
[  261s]             return resp
[  261s]
[  261s]         async def m_old2(app, handler):
[  261s]             async def middleware(request):
[  261s]                 resp = await handler(request)
[  261s]                 resp.text += "[old style 2]"
[  261s]                 return resp
[  261s]
[  261s]             return middleware
[  261s]
[  261s]         @web.middleware
[  261s]         async def m_new2(request, handler):
[  261s]             resp = await handler(request)
[  261s]             resp.text += "[new style 2]"
[  261s]             return resp
[  261s]
[  261s]         middlewares = m_old1, m_new1, m_old2, m_new2
[  261s]
[  261s]         with pytest.warns(DeprecationWarning) as w:
[  261s]             app = web.Application(middlewares=middlewares)
[  261s]             app.router.add_route("GET", "/", handler)
[  261s]             client = await aiohttp_client(app)
[  261s]             resp = await client.get("/")
[  261s]             assert 200 == resp.status
[  261s]             txt = await resp.text()
[  261s]             assert "OK[new style 2][old style 2][new style 1][old style 1]" == txt
[  261s]
[  261s] >       assert len(w) == 2
[  261s] E       assert 3 == 2
[  261s] E         +3
[  261s] E         -2
[  261s]
[  261s] tests/test_web_middleware.py:467: AssertionError
[  261s] ___________________ test_old_style_middleware_class[pyloop] ____________________
[  261s]
[  261s] loop = <_UnixSelectorEventLoop running=False closed=False debug=False>
[  261s] aiohttp_client = <function aiohttp_client.<locals>.go at 0x7fde94fd4dc0>
[  261s]
[  261s]     async def test_old_style_middleware_class(loop, aiohttp_client) -> None:
[  261s]         async def handler(request):
[  261s]             return web.Response(body=b"OK")
[  261s]
[  261s]         class Middleware:
[  261s]             async def __call__(self, app, handler):
[  261s]                 async def middleware(request):
[  261s]                     resp = await handler(request)
[  261s]                     assert 200 == resp.status
[  261s]                     resp.set_status(201)
[  261s]                     resp.text = resp.text + "[old style middleware]"
[  261s]                     return resp
[  261s]
[  261s]                 return middleware
[  261s]
[  261s]         with pytest.warns(DeprecationWarning) as warning_checker:
[  261s]             app = web.Application()
[  261s]             app.middlewares.append(Middleware())
[  261s]             app.router.add_route("GET", "/", handler)
[  261s]             client = await aiohttp_client(app)
[  261s]             resp = await client.get("/")
[  261s]             assert 201 == resp.status
[  261s]             txt = await resp.text()
[  261s]             assert "OK[old style middleware]" == txt
[  261s]
[  261s] >       assert len(warning_checker) == 1
[  261s] E       assert 2 == 1
[  261s] E         +2
[  261s] E         -1
[  261s]
[  261s] tests/test_web_middleware.py:506: AssertionError
[  261s] ___________________ test_new_style_middleware_class[pyloop] ____________________
[  261s]
[  261s] loop = <_UnixSelectorEventLoop running=False closed=False debug=False>
[  261s] aiohttp_client = <function aiohttp_client.<locals>.go at 0x7fde94fe3940>
[  261s]
[  261s]     async def test_new_style_middleware_class(loop, aiohttp_client) -> None:
[  261s]         async def handler(request):
[  261s]             return web.Response(body=b"OK")
[  261s]
[  261s]         @web.middleware
[  261s]         class Middleware:
[  261s]             async def __call__(self, request, handler):
[  261s]                 resp = await handler(request)
[  261s]                 assert 200 == resp.status
[  261s]                 resp.set_status(201)
[  261s]                 resp.text = resp.text + "[new style middleware]"
[  261s]                 return resp
[  261s]
[  261s]         with pytest.warns(None) as warning_checker:
[  261s]             app = web.Application()
[  261s]             app.middlewares.append(Middleware())
[  261s]             app.router.add_route("GET", "/", handler)
[  261s]             client = await aiohttp_client(app)
[  261s]             resp = await client.get("/")
[  261s]             assert 201 == resp.status
[  261s]             txt = await resp.text()
[  261s]             assert "OK[new style middleware]" == txt
[  261s]
[  261s] >       assert len(warning_checker) == 0
[  261s] E       assert 1 == 0
[  261s] E         +1
[  261s] E         -0
[  261s]
[  261s] tests/test_web_middleware.py:540: AssertionError
[  261s] ___________________ test_new_style_middleware_method[pyloop] ___________________
[  261s]
[  261s] loop = <_UnixSelectorEventLoop running=False closed=False debug=False>
[  261s] aiohttp_client = <function aiohttp_client.<locals>.go at 0x7fde94fd4040>
[  261s]
[  261s]     async def test_new_style_middleware_method(loop, aiohttp_client) -> None:
[  261s]         async def handler(request):
[  261s]             return web.Response(body=b"OK")
[  261s]
[  261s]         class Middleware:
[  261s]             @web.middleware
[  261s]             async def call(self, request, handler):
[  261s]                 resp = await handler(request)
[  261s]                 assert 200 == resp.status
[  261s]                 resp.set_status(201)
[  261s]                 resp.text = resp.text + "[new style middleware]"
[  261s]                 return resp
[  261s]
[  261s]         with pytest.warns(None) as warning_checker:
[  261s]             app = web.Application()
[  261s]             app.middlewares.append(Middleware().call)
[  261s]             app.router.add_route("GET", "/", handler)
[  261s]             client = await aiohttp_client(app)
[  261s]             resp = await client.get("/")
[  261s]             assert 201 == resp.status
[  261s]             txt = await resp.text()
[  261s]             assert "OK[new style middleware]" == txt
[  261s]
[  261s] >       assert len(warning_checker) == 0
[  261s] E       assert 1 == 0
[  261s] E         +1
[  261s] E         -0
[  261s]
[  261s] tests/test_web_middleware.py:566: AssertionError
[  261s] =============================== warnings summary ===============================
[  261s] tests/test_client_session.py: 10 warnings
[  261s]   /home/abuild/rpmbuild/BUILDROOT/python-aiohttp-3.7.4-0.x86_64/usr/lib64/python3.9/site-packages/aiohttp/connector.py:83: DeprecationWarning: Connector.close() is a coroutine, please use await connector.close()
[  261s]     warnings.warn(
[  261s]
[  261s] tests/test_client_session.py::test_connector_loop[pyloop]
[  261s]   /usr/lib/python3.9/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: <function ClientSession.__del__ at 0x7fde982f7af0>
[  261s]
[  261s]   Traceback (most recent call last):
[  261s]     File "/home/abuild/rpmbuild/BUILDROOT/python-aiohttp-3.7.4-0.x86_64/usr/lib64/python3.9/site-packages/aiohttp/client.py", line 326, in __del__
[  261s]       if not self.closed:
[  261s]     File "/home/abuild/rpmbuild/BUILDROOT/python-aiohttp-3.7.4-0.x86_64/usr/lib64/python3.9/site-packages/aiohttp/client.py", line 963, in closed
[  261s]       return self._connector is None or self._connector.closed
[  261s]   AttributeError: 'ClientSession' object has no attribute '_connector'
[  261s]
[  261s]     warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
[  261s]
[  261s] tests/test_loop.py: 2 warnings
[  261s] tests/test_test_utils.py: 13 warnings
[  261s]   /usr/lib64/python3.9/asyncio/base_events.py:1518: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
[  261s]     await tasks.sleep(0, loop=self)
[  261s]
[  261s] tests/test_run_app.py: 46 warnings
[  261s]   /home/abuild/rpmbuild/BUILDROOT/python-aiohttp-3.7.4-0.x86_64/usr/lib64/python3.9/site-packages/aiohttp/web.py:444: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
[  261s]     asyncio.gather(*to_cancel, loop=loop, return_exceptions=True)
[  261s]
[  261s] -- Docs: https://docs.pytest.org/en/stable/warnings.html
[  261s] =========================== short test summary info ============================
[  261s] SKIPPED [1] tests/test_client_session.py:672: The check is applied in DEBUG mode only
[  261s] SKIPPED [1] tests/test_connector.py:1932: Proactor Event loop present only in Windows
[  261s] SKIPPED [1] tests/test_connector.py:1940: Proactor Event loop present only in Windows
[  261s] SKIPPED [1] tests/test_connector.py:1951: Proactor Event loop present only in Windows
[  261s] SKIPPED [1] tests/test_connector.py:2102: Proactor Event loop present only in Windows
[  261s] SKIPPED [1] tests/test_http_parser.py:824: C based HTTP parser not available
[  261s] SKIPPED [1] tests/test_multipart_helpers.py:99: need more smart parser which respects quoted text
[  261s] SKIPPED [1] tests/test_multipart_helpers.py:446: should raise decoding error: %82 is invalid for latin1
[  261s] SKIPPED [1] tests/test_multipart_helpers.py:455: should raise decoding error: %E4 is invalid for utf-8
[  261s] SKIPPED [1] tests/test_multipart_helpers.py:510: urllib.parse.unquote is tolerate to standalone % chars
[  261s] SKIPPED [1] tests/test_multipart_helpers.py:519: urllib.parse.unquote is tolerate to standalone % chars
[  261s] SKIPPED [1] tests/test_proxy_functional.py:199: we need to reconsider how we test this
[  261s] SKIPPED [1] tests/test_proxy_functional.py:223: we need to reconsider how we test this
[  261s] SKIPPED [1] tests/test_web_app.py:297: The check is applied in DEBUG mode only
[  261s] SKIPPED [1] tests/test_web_functional.py:1927: C based HTTP parser not available
[  261s] SKIPPED [1] tests/test_web_request.py:681: The check is applied in DEBUG mode only
[  261s] SKIPPED [1] tests/test_web_runner.py:124: Proactor Event loop present only in Windows
[  261s] SKIPPED [1] tests/test_web_runner.py:134: Proactor Event loop present only in Windows
[  261s] = 5 failed, 2438 passed, 18 skipped, 4 deselected, 3 xfailed, 72 warnings in 107.37s (0:01:47) =
```

[Complete build log](https://github.com/aio-libs/aiohttp/files/7129867/_log.txt) with all versions of packages used and steps taken.

### To Reproduce

1. Run test suite

### Expected behavior

All tests should pass

### Logs/tracebacks

```python-traceback
see above
```

### Python Version

```console
[  153s] + PYTHONPATH=/home/abuild/rpmbuild/BUILDROOT/python-aiohttp-3.7.4-0.x86_64/usr/lib64/python3.9/site-packages
[  153s] + PYTHONDONTWRITEBYTECODE=1
[  153s] + pytest-3.9 --ignore=_build.python36 --ignore=_build.python39 --ignore=_build.python38 -v --ignore ./aiohttp -rs -k 'not (test_aiohttp_request_coroutine or test_mark_formdata_as_processed or test_aiohttp_plugin_async )'
[  153s] ============================= test session starts ==============================
[  153s] platform linux -- Python 3.9.7, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3.9
[  153s] cachedir: .pytest_cache
[  153s] rootdir: /home/abuild/rpmbuild/BUILD/aiohttp-3.7.4, configfile: pytest.ini
[  153s] plugins: mock-3.6.1, timeout-1.4.2
[  156s] collecting ... collected 2468 items / 4 deselected / 2464 selected
```
```


### aiohttp Version

```console
3.7.4
```

### multidict Version

```console
5.1.0
```

### yarl Version

```console
1.6.3
```

### OS

Linux/openSUSE

### Related component

Server

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

**Comment by webknjaz** [comment]
Try checking branch 3.8. Also, we need a reproducer that is not specific to your CI env.

**Comment by webknjaz** [comment]
> "{message : DeprecationWarning('The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.'), category : 'DeprecationWarning', filename : '/usr/lib64/python3.9/asyncio/base_events.py', lineno : 1518, line : None}"

Ah, it looks like newer Python versions generate more deprecation warnings. You could add this to `filterwarnings` in `pytest.ini` and ignore it. FWIW the 3.7 release stream will not get any updates. But if you're able to reproduce it under v3.8, feel free to open a new issue with a more precise STR.

**Comment by mcepl** [comment]
> Ah, it looks like newer Python versions generate more deprecation warnings. You could add this to `filterwarnings` in `pytest.ini` and ignore it

Thanks, it seems like really the difference is between 3.9.6 and 3.9.7 versions of Python. 
This [patch](https://github.com/aio-libs/aiohttp/files/7132108/python39-failures.patch.txt) makes test pass, but it is a brutal hack, some better solution should be created or these tests completely abolished.

> FWIW the 3.7 release stream will not get any updates. But if you're able to reproduce it under v3.8, feel free to open a new issue with a more precise STR.

This is nice attitude towards your users: either use unreleased development version or we will ignore you. Lovely! Anyway, that isn't the point of this issue.

**Comment by Dreamsorcerer** [comment]
> > Ah, it looks like newer Python versions generate more deprecation warnings. You could add this to `filterwarnings` in `pytest.ini` and ignore it
> 
> Thanks, it seems like really the difference is between 3.9.6 and 3.9.7 versions of Python.

Yeah, that's my fault. We found some things were broken in Python 3.10 because the tests weren't raising any deprecation warnings. I sent a PR to cpython to fix that, hence why the warnings have started appearing now in the old code.

The affected code has been updated in 3.8, so those warnings no longer occur.

> This [patch](https://github.com/aio-libs/aiohttp/files/7132108/python39-failures.patch.txt) makes test pass, but it is a brutal hack, some better solution should be created or these tests completely abolished.

The correct fix is to resolve the warnings, which are legitimate things we needed to fix, and have now been fixed in 3.8, hopefully we'll get this released soon. But, if you want to change your patch to the fixes we've done in 3.8, the diff is at: https://github.com/aio-libs/aiohttp/pull/5928/files

They are, of course, only warnings, so your existing patch is "good enough" as well.

**Comment by Dreamsorcerer** [comment]
If you're packaging the library, so users will be stuck on that version longterm, then I'd suggest possibly adding this patch as well: https://github.com/aio-libs/aiohttp/pull/5233/files#diff-2126b277e07e3fdd05e7a81da456accf24e5515a46c78c48a79db4eb166043e4

Otherwise aiohttp-session and similar will be broken for users. Unfortunately, the fix didn't make it into 3.7 in time. It's one of the reasons I'm keen to have a 3.8 release soon.

@webknjaz Is there anything I can do to help make a 3.8 release happen?

**Comment by mcepl** [comment]
@Dreamsorcerer yes, thank you, that’s very helpful. And yes, new release would be lovely as well.
